File: server/src/main/java/com/library/server/services/BookService.java
package com.library.server.services;

import com.library.server.dto.BookRequestDTO;
import com.library.server.dto.BookResponseDTO;
import com.library.server.models.Entities.Book;
import com.library.server.models.enums.BookStatus;
import com.library.server.repositories.BookRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class BookService {
    private final BookRepository bookRepository;

    public BookResponseDTO addBook(BookRequestDTO request) {
        Book book = Book.builder()
                .title(request.getTitle())
                .author(request.getAuthor())
                .availability(BookStatus.AVAILABLE)
                .build();

        Book savedBook = bookRepository.save(book);
        return mapToResponseDTO(savedBook);
    }

    public List<BookResponseDTO> getAllBooks() {
        return bookRepository.findAll()
                .stream()
                .map(this::mapToResponseDTO)
                .collect(Collectors.toList());
    }

    // FIXED: Now returns DTOs
    public List<BookResponseDTO> getAvailableBooks() {
        return bookRepository.findByAvailability(BookStatus.AVAILABLE)
                .stream()
                .map(this::mapToResponseDTO)
                .collect(Collectors.toList());
    }

    // FIXED: Now returns DTOs
    public List<BookResponseDTO> searchBooks(String query) {
        return bookRepository.findByTitleContainingIgnoreCaseOrAuthorContainingIgnoreCase(query, query)
                .stream()
                .map(this::mapToResponseDTO)
                .collect(Collectors.toList());
    }

    private BookResponseDTO mapToResponseDTO(Book book) {
        return BookResponseDTO.builder()
                .bookId(book.getBookId())
                .title(book.getTitle())
                .author(book.getAuthor())
                .availability(book.getAvailability())
                .build();
    }
}

File: frontend/src/services/bookService.js
import API from "./api";

// GET ALL BOOKS

export const getAllBooks = async () => {
  const response = await API.get("/books");

  return response.data.data;
};

// GET AVAILABLE BOOKS

export const getAvailableBooks = async () => {
  const response = await API.get(
    "/books/available"
  );

  return response.data.data;
};

// ADD BOOK

export const addBook = async (
  bookData
) => {
  const response = await API.post(
    "/books",
    bookData
  );

  return response.data.data;
};

// SEARCH BOOKS

export const searchBooks = async (
  query
) => {
  const response = await API.get(
    `/books/search?query=${query}`
  );

  return response.data.data;
};